packages/python/plotly/plotly/express/_core.py

```diff
@@ -754,6 +754,215 @@ def apply_default_cascade(args):
         args["marginal_x"] = None
 
 
+def _name_heuristic(argument, field_name, used_col_names):
+    if isinstance(argument, int):
+        argument = str(argument)
+    if field_name not in used_col_names:
+        return field_name
+    else:
+        raise NameError(
+            "A name conflict was encountered for argument %s. "
+            "A column with name %s is already used." % (field_name, field_name)
+        )
+
+
+def _initialize_argument_col_names(args, attrables, array_attrables):
+    """
+    This function builds a list of columns of the data_frame argument used
+    as arguments, either as str/int arguments or given as columns
+    (pandas series type).
+    """
+    df = args["data_frame"]
+    used_col_names = set()
+    for field in args:
+        if field not in attrables:
+            continue
+        names = args[field] if field in array_attrables else [args[field]]
+        if names is None:
+            continue
+        for arg in names:
+            if arg is None:
+                continue
+            if isinstance(arg, str):
+                used_col_names.add(arg)
+            if isinstance(arg, pd.core.series.Series):
+                arg_name = arg.name
+                if arg_name and hasattr(df, arg_name):
+                    in_df = arg is df[arg_name]
+                    if in_df:
+                        used_col_names.add(arg_name)
+
+    return used_col_names
+
+
+def build_dataframe(args, attrables, array_attrables):
+    """
+    Constructs a dataframe and modifies `args` in-place.
+
+    The argument values in `args` can be either strings corresponding to
+    existing columns of a dataframe, or data arrays (lists, numpy arrays,
+    pandas columns, series).
+
+    Parameters
+    ----------
+    args : OrderedDict
+        arguments passed to the px function and subsequently modified
+    attrables : list
+        list of keys into `args`, all of whose corresponding values are
+        converted into columns of a dataframe.
+    array_attrables : list
+        argument names corresponding to iterables, such as `hover_data`, ...
+    """
+    # Cast data_frame argument to DataFrame (it could be a numpy array, dict etc.)
+    df_provided = args["data_frame"] is not None
+    if df_provided and not isinstance(args["data_frame"], pd.DataFrame):
+        args["data_frame"] = pd.DataFrame(args["data_frame"])
+
+    # We start from an empty DataFrame
+    df = pd.DataFrame()
+
+    # Initialize set of column names
+    # These are reserved names
+    if df_provided:
+        reserved_names = _initialize_argument_col_names(
```

Comment:
consistency-wise, maybe we should rename this function `_get_reserved_col_names` or something?